stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -14,434 +14,241 @@
-    const Op_local_slice_t slice(op_.begin() + start, op_.begin() + end);
-    return slice;
-  }
-
-  inline std::size_t num_elements() const { return 0; }
-
-  inline void build(vari** varis, double* partials) const {}
-};
+    // Works with anything that has operator[Integral] defined
+    template <typename... Pargs, typename Vec,
+              require_vector_like_vt<is_var, Vec>...>
+    void accumulate_adjoints(double* dest, const Vec& x, const Pargs&... args) {
+      for (size_t i = 0; i < x.size(); ++i) {
+        dest[i] += x[i].adj();
+      }
+      accumulate_adjoints(dest + x.size(), args...);
+    }
 
-/*
-template <typename T, typename = void>
-struct local_operand_and_partials<T, > {
-};
-*/
-
-template <class ReduceFunction, class M, class T, class Arg1, class Arg2,
-          class Arg3, class Arg4>
-struct reduce_sum_impl<ReduceFunction, M, T, Arg1, Arg2, Arg3, Arg4, var> {
-  using vmapped_t = std::vector<M>;
-  using arg1_t = std::vector<Arg1>;
-  using arg2_t = std::vector<Arg2>;
-  using arg3_t = std::vector<Arg3>;
-  using arg4_t = std::vector<Arg4>;
-
-  using vmapped_op_t = local_operand_and_partials<M>;
-  using arg1_local_op_t = local_operand_and_partials<Arg1>;
-  using arg2_local_op_t = local_operand_and_partials<Arg2>;
-  using arg3_local_op_t = local_operand_and_partials<Arg3>;
-  using arg4_local_op_t = local_operand_and_partials<Arg4>;
+    // Works on anything with a operator()
+    template <typename... Pargs, typename Mat,
+              require_t<is_detected<Mat, operator_paren_access_t>>...,
+              require_t<is_var<value_type_t<Mat>>>...>
+    void accumulate_adjoints(double* dest, const Mat& x, const Pargs&... args) {
+      for (size_t i = 0; i < x.size(); ++i) {
+        dest[i] += x(i).adj();
+      }
+      accumulate_adjoints(dest + x.size(), args...);
+    }
 
-  struct recursive_reducer {
-    vmapped_op_t& op_vmapped_;
-    arg1_local_op_t op_arg1_;
-    arg2_local_op_t op_arg2_;
-    arg3_local_op_t op_arg3_;
-    arg4_local_op_t op_arg4_;
-
-    const std::size_t num_vmapped_adjoints_per_element_;
-    double terms_sum_;
-
-    recursive_reducer(vmapped_op_t& op_vmapped,
-                      std::size_t num_vmapped_adjoints_per_element,
-                      const T& init, const arg1_t& arg1, const arg2_t& arg2,
-                      const arg3_t& arg3, const arg4_t& arg4)
-        : op_vmapped_(op_vmapped),
-          op_arg1_(arg1),
-          op_arg2_(arg2),
-          op_arg3_(arg3),
-          op_arg4_(arg4),
-          num_vmapped_adjoints_per_element_(num_vmapped_adjoints_per_element),
-          terms_sum_(as_value(init)) {}
+    // Anything with a scalar type of Arithmetic gets tossed
+    template <typename Arith, require_arithmetic_t<scalar_type_t<Arith>>...,
+              typename... Pargs>
+    void accumulate_adjoints(double* dest, Arith&& x, const Pargs&... args) {
+      accumulate_adjoints(dest, args...);
+    }
```

Comment:
Ben I changed this mostly because I don't like the idea of having one huge catchall function. Would rather specialize for things we know

Reply:
I agree. I would especially like to avoid the catch-all deep copy thing here: ```var deep_copy(const var& arg) { return var(arg.val()); }```. That should only catch Arithmetic types as well. Arithmetric means non-autodiff, right?

Reply:
Yeah it just calls `std::arithmetic` under the hood to enable the method so only non-autodiff types

https://en.cppreference.com/w/cpp/types/is_arithmetic